src/NServiceBus.Core/Performance/Metrics/ReceiveDiagnosticsBehavior.cs

```diff
@@ -0,0 +1,46 @@
+namespace NServiceBus.Performance.Metrics
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Metrics;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class ReceiveDiagnosticsBehavior : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
```

Comment:
have you considered using the notifications rather than a behavior to raise the metrics? that way the state (success/failure) would be more reliable since the notifications are raised after the fact. Haven't thought too much about this but it might be an alternative to consider.

Reply:
Checked how the Notification is used in NServiceBus.Metrics. It is used in addition to a behavior, to report on immediate and delayed retries. In fact, in that package a behavior is used, similar to what I did here.